streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java

```diff
@@ -215,14 +216,18 @@ synchronized void unlockGlobalState() throws IOException {
     synchronized void unlock(final TaskId taskId) throws IOException {
         final LockAndOwner lockAndOwner = locks.get(taskId);
         if (lockAndOwner != null && lockAndOwner.owningThread.equals(Thread.currentThread().getName())) {
-            locks.remove(taskId);
-            lockAndOwner.lock.release();
+            try {
+                lockAndOwner.lock.release();
+            } catch (final ClosedChannelException cce) {
+                log.warn("{} Channel closed unexpectedly before lock release.", logPrefix(), cce);
+            }
             log.debug("{} Released state dir lock for task {}", logPrefix(), taskId);
 
             final FileChannel fileChannel = channels.remove(taskId);
             if (fileChannel != null) {
                 fileChannel.close();
```

Comment:
Should we catch `IOExepction` here, too, and make sure we remove the `taskId` from `locks` before we rethrow it?

Reply:
I'm not sure; this is the root of my general un-easiness with this change.

If I understand the API correctly, when we release the lock and close the channel, we are removing the lockfile. If we get an IOException, it *might* be because the file is already gone, but it might also be any number of other causes.

If we stop tracking the lock before we are sure the lockfile is gone, don't we risk a kind of deadlock in which the lockfile is there, but no one thinks they own it, and therefore no one will release it?